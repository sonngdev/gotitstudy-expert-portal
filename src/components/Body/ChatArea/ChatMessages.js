import React from 'react';
import { useSelector } from 'react-redux';
import { BubbleChat } from '@gotitinc/design-system';
import { chatMessagesSelector } from '../../../redux/selectors';
import './styles/ChatMessages.css';

export function ChatMessages() {
  const messages = useSelector(chatMessagesSelector);
  const messageBubbles = messages.map(({ type, avatar, text }, i) => (
    <BubbleChat type={type} avatar={avatar} text={text} key={i} />
  ));

  return (
    <div className="chat-messages">
      {messageBubbles}
    </div>
  );
}
