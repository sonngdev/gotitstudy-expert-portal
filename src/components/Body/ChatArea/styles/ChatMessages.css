.chat-messages {
  height: calc(100vh - 160px);
  overflow: auto;
  margin-bottom: 10px;
}
