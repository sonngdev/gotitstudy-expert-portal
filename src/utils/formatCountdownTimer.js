export function formatCountdownTimer(milisecond) {
  const minute = Math.floor(milisecond / 1000 / 60);
  const second = Math.floor(milisecond / 1000) - 60 * minute;

  return [minute, second].map((n) => String(n).padStart(2, '0')).join(':');
}
