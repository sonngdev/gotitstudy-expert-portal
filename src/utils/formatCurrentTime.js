import moment from 'moment-timezone';
import { DEFAULT_TIMEZONE, CURRENT_TIME_FORMAT } from '../constants';

export function formatCurrentTime() {
  return moment().tz(DEFAULT_TIMEZONE).format(CURRENT_TIME_FORMAT);
}
