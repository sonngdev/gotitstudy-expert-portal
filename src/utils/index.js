export { getCurrentTime, formatCountdownTimer, formatCurrentTime } from './time';
export { useChatMessagesList, useChatMessagesAdd } from './useChatMessages';
export { useScrollToBottom } from './useScrollToBottom';
export { useTinyMceEditor } from './useTinyMceEditor';
