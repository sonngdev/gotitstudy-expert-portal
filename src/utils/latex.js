/* eslint-disable prefer-arrow-callback */

tinymce.PluginManager.add('latex', function addLatexPlugin(editor) {
  /**
  |--------------------------------------------------
  | VARIABLES
  |--------------------------------------------------
  */
  const latexClass = 'Latexformula';

  const latexPreviewId = `${latexClass}-preview`;


  /**
  |--------------------------------------------------
  | HELPERS
  |--------------------------------------------------
  */
  const isLatexFormula = (node) => [...node.classList].includes(latexClass);

  const getExpression = (dialog) => dialog.getData().expression.trim();

  const getSelectedExpression = () => {
    const { selection } = editor;
    const node = selection && selection.getNode();
    return node && isLatexFormula(node) ? node.alt : '';
  };

  const getFormulaSrc = (expression) => (
    /**
     * LaTeX Math as SVG image:
     * https://math.now.sh/home
     */
    `https://math.now.sh?from=${encodeURIComponent(expression)}`
  );

  const setPreview = (content) => {
    document.getElementById(latexPreviewId).innerHTML = content;
  };


  /**
  |--------------------------------------------------
  | DIALOG
  |--------------------------------------------------
  */
  const openDialog = () => editor.windowManager.open({

    title: 'LaTeX expression',

    body: {
      type: 'panel',
      items: [
        { type: 'textarea', name: 'expression', label: 'Expression' },
        { type: 'button', name: 'preview', text: 'Preview' },
        { type: 'htmlpanel', html: `<div id="${latexPreviewId}"></div>` },
      ],
    },

    buttons: [
      { type: 'cancel', text: 'Cancel' },
      { type: 'submit', text: 'Insert', primary: true },
    ],

    initialData: {
      expression: getSelectedExpression(),
    },

    onAction(dialog, details) {
      if (details.name !== 'preview') return;

      const expression = getExpression(dialog);
      if (!expression) return;

      const src = getFormulaSrc(expression);
      const content = `<img src="${src}" alt="${expression}" role="math" />`;
      setPreview(content);
    },

    onSubmit(dialog) {
      const expression = getExpression(dialog);
      if (!expression) return;

      const src = getFormulaSrc(expression);
      const content = `<img class="${latexClass}" src="${src}" alt="${expression}" role="math"/>`;
      editor.insertContent(content);

      dialog.close();
    },
  });


  /**
  |--------------------------------------------------
  | REGISTER
  |--------------------------------------------------
  */
  editor.ui.registry.addButton('latex', {
    text: 'LaTeX',
    onAction: openDialog,
  });

  editor.ui.registry.addMenuItem('latex', {
    text: 'LaTeX',
    onAction: openDialog,
  });

  editor.on('DblClick', ({ target }) => {
    if (isLatexFormula(target)) openDialog();
  });

  return {
    getMetadata: () => ({ name: 'LaTeX plugin for tinymce 5' }),
  };
});
