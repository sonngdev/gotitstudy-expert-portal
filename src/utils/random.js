/**
 * For trivial randomness, this solution is sufficient
 * https://stackoverflow.com/questions/1349404/generate-random-string-characters-in-javascript
 */
export function weakRandomString(len) {
  if (len >= 10) throw new Error('String length must be smaller than 10');
  return Math.random().toString(36).substring(2, 2 + len);
}
