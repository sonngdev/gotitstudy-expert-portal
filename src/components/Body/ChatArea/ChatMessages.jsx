import React from 'react';
import { useSelector } from 'react-redux';
import { BubbleChat } from '@gotitinc/design-system';
import './styles/ChatMessages.css';

export function ChatMessages() {
  const messages = useSelector((state) => state.chatMessages);
  const messageBubbles = messages.map(({ type, avatar, text }, i) => (
    <BubbleChat type={type} avatar={avatar} text={text} key={i} />
  ));

  return (
    <div className="chat-messages">
      {messageBubbles}
    </div>
  );
}
