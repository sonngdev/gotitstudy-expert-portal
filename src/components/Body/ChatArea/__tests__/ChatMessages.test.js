import React from 'react';
import { shallow } from 'enzyme';
import { useChatMessages, useScrollToBottom } from '../../../../utils';
import { ChatMessages } from '../ChatMessages';

jest.mock('../../../../utils', () => ({
  __esModule: true,
  useChatMessages: jest.fn(() => [
    {
      type: 'reviser',
      avatar: 'asker',
      text: 'How do you do this?',
    },
    {
      type: 'system',
      avatar: 'gotit',
      text: 'This is a system message from Got It!.',
    },
    {
      type: 'sender',
      avatar: 'expert',
      text: 'Let me take a look.',
    },
  ]),
  useScrollToBottom: jest.fn(),
}));

describe('<ChatMessages />', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<ChatMessages />);
  });

  it('get messages from custom hook', () => {
    expect(useChatMessages).toHaveBeenCalled();
  });

  it('scrolls to bottom after every render', () => {
    expect(useScrollToBottom).toHaveBeenCalled();
  });

  it('renders all messages in the conversation in order', () => {
    const chats = wrapper.find('BubbleChat');
    expect(chats.length).toBe(3);
    expect(chats.at(0).prop('type')).toBe('reviser');
    expect(chats.at(2).prop('type')).toBe('sender');
  });
});
