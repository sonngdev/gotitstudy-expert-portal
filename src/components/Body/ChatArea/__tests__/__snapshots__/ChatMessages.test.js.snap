// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChatMessages /> renders 1`] = `
"<div className=\\"chat-messages u-paddingRightSmall\\">
  <BubbleChat type=\\"reviser\\" avatar=\\"asker\\" text=\\"How do you do this?\\" />
  <BubbleChat type=\\"system\\" avatar=\\"gotit\\" text=\\"This is a system message from Got It!.\\" />
  <BubbleChat type=\\"sender\\" avatar=\\"expert\\" text=\\"Let me take a look.\\" />
</div>"
`;
